ui/client/src/views/Keys.tsx

```diff
@@ -0,0 +1,368 @@
+  }
+
+  const handleSortChange = (column: string) => {
+    if (column === sortBy) {
+      const order = sortOrder === 'asc' ? 'desc' : 'asc';
+      setSortOrder(order);
+      window.localStorage.setItem(constants.KEYS_SORT_ORDER_STORAGE_KEY, order);
+    } else {
+      window.localStorage.setItem(constants.KEYS_SORT_BY_STORAGE_KEY, column);
+      if (sortOrder !== 'asc') {
+        window.localStorage.setItem(constants.KEYS_SORT_ORDER_STORAGE_KEY, 'asc');
+        setSortOrder('asc');
+      }
+      setSortBy(column);
+    }
+  };
+
+  let breadcrumbContent: JSX.Element[] = [];
+  if (parent !== '') {
+    const segments = parent.split('.');
+    let segmentStack: string[] = [];
+    for (const segment of segments) {
+      segmentStack.push(segment);
+      const target = segmentStack.join('.');
+      breadcrumbContent.push(
+        <Link underline="none"
+          key={segment}
+          href=""
+          sx={{ textTransform: 'none' }}
+          onClick={event => {
+            event.preventDefault();
+            setFilter(undefined);
+            setParent(target);
+          }}>
+          {segment === '' ? t('root') : segment}
+        </Link>
+      )
+    }
+  }
+  if (filter !== undefined) {
+    breadcrumbContent.push(
+      <Link underline="none"
+        key={filter}
+        sx={{ textTransform: 'none' }}>
+        {filter}
+      </Link>);
+  }
+
+  const getEthAddress = (key: IKeyEntry) => {
+    const entry = key.verifiers?.find(entry => entry.type === 'eth_address');
+    if (entry !== undefined) {
+      return <Hash
+        Icon={<Captions size="18px" />}
+        title={entry.algorithm}
+        hash={entry.verifier}
+        hideTitle />
+    }
+    return <RemoveIcon color="disabled" />;
+  };
+
+  const getOtherVerifiers = (key: IKeyEntry) => {
+    if (key.verifiers !== null) {
+      const entries = key.verifiers.filter(entry => entry.type !== 'eth_address');
+      if (entries.length === 1) {
+        return <Hash
+          Icon={<Signature size="18px" />}
+          title={entries[0].algorithm}
+          hash={entries[0].verifier}
+          hideTitle />
+      } else if (entries.length > 1) {
+        return (
+          <Button
+            variant="contained"
+            disableElevation
+            color="primary"
+            size="small"
+            sx={{ minWidth: '110px', paddingTop: 0, paddingBottom: 0, textTransform: 'none', fontWeight: '400', whiteSpace: 'nowrap' }}
+            onClick={() => { setSelectedVerifiers(entries); setVerifiersDialogOpen(true) }}
+          >
+            {t('manyN', { n: entries.length })}
+          </Button>
+        );
+      }
+    }
+    return <RemoveIcon color="disabled" />;
+  };
+
+  const handleChangePage = (
+    _event: React.MouseEvent<HTMLButtonElement> | null,
+    newPage: number
+  ) => {
+    if (newPage === 0) {
+      setRefEntries([]);
+    } else if (newPage > page) {
+      refEntries.push(keys[keys.length - 1]);
```

Comment:
So the depth of the stack here will be the number of pages forwards you've gone

Reply:
That is exactly right. This auxiliary stack structure is kept in memory so that we can go back 1 page at a time. Going to the first page simply consists of clearing the stack.